lua/treesitter-context.lua

```diff
@@ -470,16 +470,32 @@ local function get_parent_matches(max_lines)
 end
 
 local function throttle_fn(fn)
-  local running = false
-  return function()
-    if not running then
-      running = true
-      vim.defer_fn(function()
-        fn()
-        running = false
-      end, 50)
+  local recalc_after_cooldown = false
+  local cooling_down = false
+  local function wrapped()
+    if cooling_down then
+      recalc_after_cooldown = true
+    else
+      local start = vim.loop.hrtime()
+      fn()
+      local elapsed_ms = math.floor((vim.loop.hrtime() - start) / 1e6)
+      -- If this took < 5ms, we don't need a cooldown period. This prevents the context floats from flickering
+      if elapsed_ms > 5 then
```

Comment:
Why not cool down unconditionally?

I tried this locally and I didn't notice any flickering:

```lua
local function throttle_fn(fn)
  local scheduled = false
  local cooling = false
  local function wrapped()
    if cooling then
      scheduled = true
      return
    end

    cooling = true
    fn()

    vim.defer_fn(function()
      cooling = false
      if scheduled then
        scheduled = false
        wrapped()
      end
    end, 500)
  end
  return wrapped
end
```

 I also think the cooldown period can be larger. I tried 500 without issues, but 50 (like it was) should be ok.

Reply:
Good question! I did try that at first, but cooling down unconditionally means that there is a window of time post-render where the context will not rerender. This means that it will solve the latency problem if the context float needs to be updated as a result of the _first_ scroll action, but something like `jj` for example could still see lag.

PR version on left, code from your comment on the right
https://user-images.githubusercontent.com/506791/195635810-739cf618-2a8e-4fdf-ab6b-1295913bdcb4.mp4

As you mentioned, this is using a cooldown of 500ms which would accentuate the differences here, but even with a value of 50 we would expect to see the same latency issues, just slightly less.

Reply:
Ok, how about we reduce the 5ms to 2ms, but increase the cool down to 20ms?

Reply:
Sorry, to clarify: you're suggesting taking the existing PR logic and replacing the 5ms threshold for throttling with a 2ms threshold, and replacing the `cooldown_period = 2 * elapsed_ms` with a flat 20ms, is that correct?

Reply:
Yes

Reply:
The main reason I went with a dynamic throttle time was because I was worried that in the absolute worst cases (blocking the main thread for hundreds of ms) a small constant throttle would not be enough to preserve scroll performance. Here's the comparison of the two running on ~170k lines of ruby:

https://user-images.githubusercontent.com/506791/195743959-08fc149d-683e-4995-9547-5168e527adae.mp4

Nearly identical to my eyes. I checked the responsiveness and it also looks good to me.

Reply:
Put up a commit with the new values you requested